tokio/src/task/task_set.rs

```diff
@@ -0,0 +1,224 @@
+///     let mut seen = [false; 10];
+///     while let Some(res) = set.join_one().await.unwrap() {
+///         seen[res] = true;
+///     }
+///
+///     for i in 0..10 {
+///         assert!(seen[i]);
+///     }
+/// }
+/// ```
+pub struct TaskSet<T> {
+    inner: IdleNotifiedSet<Option<JoinHandle<T>>>,
+}
+
+impl<T> TaskSet<T> {
+    /// Create a new `TaskSet`.
+    pub fn new() -> Self {
+        Self {
+            inner: IdleNotifiedSet::new(),
+        }
+    }
+
+    /// Returns the number of tasks currently in the `TaskSet`.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns whether the `TaskSet` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.inner.len() == 0
+    }
+}
+
+impl<T: Send + 'static> TaskSet<T> {
+    /// Spawn the provided task on the task set.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if called outside of a Tokio runtime.
+    pub fn spawn<F>(&mut self, task: F)
+    where
+        F: Future<Output = T>,
+        F: Send + 'static,
+    {
+        self.insert(crate::spawn(task));
+    }
+
+    /// Spawn the provided task on the provided runtime and store it in this `TaskSet`.
+    pub fn spawn_on<F>(&mut self, task: F, handle: &Handle)
+    where
+        F: Future<Output = T>,
+        F: Send + 'static,
+    {
+        self.insert(handle.spawn(task));
+    }
+
+    /// Spawn the provided task on the current [`LocalSet`] and store it in this `TaskSet`.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if it is called outside of a `LocalSet`.
+    ///
+    /// [`LocalSet`]: crate::task::LocalSet
+    pub fn spawn_local<F>(&mut self, task: F)
+    where
+        F: Future<Output = T>,
+        F: 'static,
+    {
+        self.insert(crate::task::spawn_local(task));
+    }
+
+    /// Spawn the provided task on the provided [`LocalSet`] and store it in this `TaskSet`.
+    ///
+    /// [`LocalSet`]: crate::task::LocalSet
+    pub fn spawn_local_on<F>(&mut self, task: F, local_set: &LocalSet)
+    where
+        F: Future<Output = T>,
+        F: 'static,
+    {
+        self.insert(local_set.spawn_local(task));
+    }
+
+    fn insert(&mut self, jh: JoinHandle<T>) {
+        let entry = self.inner.insert_idle(Some(jh));
+
+        // Set the waker that is notified when the task completes.
+        self.inner.with_entry_value(&entry, |jh| {
+            let jh = jh.as_mut().unwrap();
+            entry.with_context(|ctx| jh.set_join_waker(ctx.waker()))
+        });
+    }
+
+    /// Wait until one of the tasks in the set completes and returns its output.
+    ///
+    /// Returns `None` if the set is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
```

Comment:
I would probably keep this as `Option<Result<...>>`.

Reply:
Well, it's a pain to use in a `while let` loop if you do that though.

Reply:
Is it worse? It seems like returning `Option` means you can loop until all tasks are finished, while as `Result` makes it easier for one task error to blow up the loop.

```rust
while let Some(result) = set.join_one().await {

}
// or
while let Ok(Some(val)) = set.join_one().await {
    // loop canceled on first `Err`?
}
```

The `Option` _does_ make it more annoying if you want to use `?` (like `set.join_one().await?`) in the condition. I have an opinion otherwise about which is better or worse.

Reply:
I was mostly thinking of the `while let Some(result) = set.join_one().await?` use-case.

Reply:
I think I agree w/ @Darksonn but what is the precedence in `std`? I'm not sure what the best way to search is. I can't think of any methods that returns `Result<Option<_>, _>` and the only methods that return `Option<Result<...>>` would be `Iterator`, which doesn't really count.

Reply:
I do not think that std has precedence on this.